File: class/sorting/quicksort.cpp
#include <iostream>
#include <algorithm>
#include <utility>

using namespace std;

int partition(int* array, const int left, const int right){
  const int mid = left + (right - left) /2 ;
  // 1. Find a pivot value;
  const int pivot = array[mid];
  // Move the mid point value to the front
  // Swap the value to the front of the sequence,
  swap(array[mid], array[left]);
  int i = left + 1;
  int j = right;

  while(i <= j){
    // Move all values less than or equal to the pivot value to the left
    while(i <= j && array[i] <= pivot){
      i++;
    }
    // Move all values larger than the pivot to the right
    while(i <= j && array[j] > pivot){
      j--;
    }

    if(i < j){
      swap(array[i], array[j]);
    }
  }

  // Insert the pivot value at the right place.
  swap(array[i - 1], array[left]);
  return i - 1;
}

void quicksort(int* array, const int left, const int right, const int size){
  if(left >= right){
    return;
  }
  int part = partition(array, left, right);

  // Part is now at the right place, so you never have to include this pivot value in
  // sorting again.
  if(left < part - 1){
    quicksort(array, left, part - 1, size);
  }
  if(part + 1 < right){
    quicksort(array, part + 1, right, size);
  }
}

int main(){
  int array[8] = {110, 5, 10, 3, 22, 100, 1, 23};
  int sz = sizeof(array)/sizeof(array[0]);
  quicksort(array, 0, sz -1 , sz);
  for(int i = 0; i < sz; i++){
    cout << array[i] << endl;
  }
  return 0;
}

File: class/sorting/quick_sort.py
def quickSort(array, L, R):
    if(R < L + 2):
        return
    p = Partition(array, L, R)
    quickSort(array, L, p)
    quickSort(array, p, R)

def Partition(array, L, R):
    i=j=L
    pivot = R
    while i < pivot:
        if(array[i] < array[pivot]):
            swap(array, i, j)
            j+=1
        i+=1
        print("array", array)
    swap(array, pivot, j)

    return j
def swap(array, i, j):
    temp = array[i]
    array[i] = array[j]
    array[j] = temp

# Driver code to test above
arr = [10, 7, 8, 9, 1, 5]
n = len(arr)
quickSort(arr,0,n-1)
print ("Sorted array is:")
for i in range(n):
    print ("%d" %arr[i]),
